doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+
+2. All static cxx containers are still owner of all their data. This means the
+   memory which the stack allocator manages must be stored inside the cxx
+   container class.
+
+3. A function developer should not have to specify the allocator type of a
+   container. This means the allocator type is independent of the container type.
+   This has to work:
+   ```cpp
+   void myFunkyFunction(const cxx::vector<int> & bla) {}
+
+   cxx::vector<int, 20> a;
+   cxx::vector<int, 30> b;
+   cxx::vector<int> c(Allocator::heap);
+
+   myFunkyFunction(a);
+   myFunkyFunction(b);
+   myFunkyFunction(c);
+   ```
+
+4. All containers must be shared memory compatible, this means no vtables and
+   virtual. Furthermore the usage of function pointers, `cxx::function` and
+   `cxx::function_ref` is forbidden in all containers and their underlying 
+   constructs.
+
+5. Before using any cxx container the user as to call `reserve()` once otherwise
+   the container will have a default capacity of zero. It is not allowed to
+   call `reserve()` multiple times since this could violate the zero copy
+   guarantee.
+   Stack based containers do not require a `reserve()` call beforehand but
+   should provide this call as well for compatibility reasons.
+
+   * It is not possible to provide solely the capacity (`reserve` parameter) via
+     the constructor of the container since frameworks like ROS2 cannot forward
+     such arguments when for instance `loan` was called. But it is possible to
+     provide it as an alternative.
+
+6. The allocator type should not change the underlying container type, e.g. the
+   allocator should not be provided as template argument, otherwise
+   it becomes impossible to reuse defined structures in a non shared memory
+   context. This is for instance required when a user would like to store some
+   received data locally in a cache for later use.
+
+7. The allocator should not be part of copy and move-assignment operations.
+   The container should behave as closely as possible to an STL container with
+   custom allocator.
+   ```cpp
+   cxx::vector a(HeapAllocator);
+   cxx::vector b(SharedMemoryAllocator);
+   cxx::vector c(AnotherAllocator);
+   cxx::vector d(HeapAllocator);
+
+   a = b; // the content is copied from shared memory into the heap
+   a = std::move(c); // a fake move is performed and the data is again copied
+                     // into the heap
+   a = std::move(d); // real move is performed
+   ```
+   After a copy- or move assignment operation the allocator which was provided
+   in the constructor will not change.
+
+   When the copy- or move-constructor is used the allocator of the origin shall
+   be used in the newly created container.
+
+   * Real copy- and move operations could have some unwanted side effects. Lets
+     assume one would like to copy/move a shared memory located vector into a local
+     cache. If the allocator would change, the container has to allocate another
```

Comment:
Can/should the allocator of an individual vector ever change once constructed? What would be the semantics?